crypto/x509/x_pubkey.c

```diff
@@ -263,9 +263,43 @@ int i2d_PUBKEY(const EVP_PKEY *a, unsigned char **pp)
     if (a->ameth != NULL && a->ameth->pub_encode != NULL
         && !a->ameth->pub_encode(xpk, a))
         goto error;
-    xpk->pkey = (EVP_PKEY *)a;
-    ret = i2d_X509_PUBKEY(xpk, pp);
-    xpk->pkey = NULL;
+    if (a->ameth != NULL) {
+        xpk->pkey = (EVP_PKEY *)a;
+        ret = i2d_X509_PUBKEY(xpk, pp);
+        xpk->pkey = NULL;
+        goto error;
+    }
+    {
+        const char *serprop = OSSL_SERIALIZER_PUBKEY_TO_DER_PQ;
+        OSSL_SERIALIZER_CTX *ctx =
+            OSSL_SERIALIZER_CTX_new_by_EVP_PKEY(a, serprop);
+        BIO *out = BIO_new(BIO_s_mem());
+
+        if (ctx != NULL
+            && out != NULL
+            && OSSL_SERIALIZER_CTX_get_serializer(ctx) != NULL
+            && OSSL_SERIALIZER_to_bio(ctx, out)) {
+            BUF_MEM *buf = NULL;
+
+            BIO_get_mem_ptr(out, &buf);
+            ret = buf->length;
+
+            if (pp != NULL) {
```

Comment:
what is the point of calling this method if pp is NULL.. That looks like an error case to me.

Reply:
Nope. And i2d function is still supposed to return the length of the would be encoding.